packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+    }
+
+    public get disposed(): boolean {
+        return this._queue.disposed;
+    }
+
+    public async pause(): Promise<void> {
+        return this._queue.pause();
+    }
+
+    public resume(): void {
+        return this._queue.resume();
+    }
+
+    public async waitTillProcessingDone(): Promise<{ count: number; duration: number; }> {
+        return this._queue.waitTillProcessingDone();
+    }
+}
+
+export class VirtualDeltaManager<TConnectionManager extends IConnectionManager>
+    extends DeltaManager<TConnectionManager> {
+
+    private _virtualSequenceNumber = 0;
+    private _clientSequenceNumber = 0;
+
+    /**
+     * Key: real client sequence number
+     * Value: virtual client sequence number
+     */
+    private readonly _clientSequenceNumberMap = new Map<number, number>();
+
+    /**
+     * TODO: removal of entries based on minimumSequenceNumber?
+     * Key: real sequence number
+     * Value: virtual sequence number
+     */
+    private readonly _sequenceNumberMap = new Map<number, number>();
+
+    private readonly _virtualizedInbound: IDeltaQueue<ISequencedDocumentMessage>;
+    private readonly _virtualizedOutbound: IDeltaQueue<IDocumentMessage[]>;
+
+    public get inbound(): IDeltaQueue<ISequencedDocumentMessage> {
+        return this._virtualizedInbound;
+    }
+
+    public get outbound(): IDeltaQueue<IDocumentMessage[]> {
+        return this._virtualizedOutbound;
+    }
+
+    constructor(
+        serviceProvider: () => IDocumentService | undefined,
+        logger: ITelemetryLogger,
+        active: () => boolean,
+        createConnectionManager: (props: IConnectionManagerFactoryArgs) => TConnectionManager,
+    ) {
+        super(serviceProvider, logger, active, createConnectionManager);
+        this._virtualizedInbound = new VirtualDeltaQueueProxy(
+            this._inbound,
+            (message: ISequencedDocumentMessage, storeValue?: boolean) => {
+                return this.virtualizeSequencedMessage(message, storeValue);
+            });
+        this._virtualizedOutbound = new VirtualDeltaQueueProxy(
+            this.connectionManager.outbound,
+            (messages: IDocumentMessage[], _storeValue?: boolean) => {
+                return messages.map((message) => {
+                    return {
+                        ...message,
+                        referenceSequenceNumber: this.virtualizeSequenceNumber(message.referenceSequenceNumber),
+                        clientSequenceNumber: this.virtualizeClientSequenceNumber(message.clientSequenceNumber),
+                    }
+                });
+            });
+    }
+
+    public async attachOpHandler(
+        minSequenceNumber: number,
+        sequenceNumber: number,
+        term: number,
+        handler: IDeltaHandlerStrategy,
+        prefetchType: "cached" | "all" | "none" = "none",
+    ) {
+        this._virtualSequenceNumber = sequenceNumber;
+
+        this._sequenceNumberMap.set(sequenceNumber, sequenceNumber);
```

Comment:
initialization seems non-complete (for prototype).
I think you want to init all the numbers in teh range of MSN...sequenceNumber, as you will observe ops with referenceSequenceNumber from other clients in between of these low & hi bounds.